### PR TITLE
Tick allocation refactor

### PR DESCRIPTION
**Pull request by TiviPlus**
TG pr is technically but this was previously broken so I need a round of testing before I PR it to TG

:cl:
refactor: refactored the MC to better compensate for lag caused by byond processes
/:cl:

### CODE REVIEW

**Comment by TiviPlus** [comment]
Merged upstream

File: code/__DEFINES/_tick.dm [hunk 1]
```diff
@@ -1,14 +1,19 @@
 /// Percentage of tick to leave for master controller to run
-#define MAPTICK_MC_MIN_RESERVE 70
+#define MAPTICK_MC_MIN_RESERVE 60
```
Comment (MrStonedOne) [inline code/__DEFINES/_tick.dm:2]: try this at something much much lower like 40. tgmc doesn't need the mc to do as much.

File: code/__DEFINES/_tick.dm [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
-#define MAPTICK_LAST_INTERNAL_TICK_USAGE ((GLOB.internal_tick_usage / world.tick_lag) * 100)
-/// Tick limit while running normally
+#define MAPTICK_LAST_INTERNAL_TICK_USAGE ((Master.normalized_internal_tick_usage / world.tick_lag) * 100)
+/// Amount of a tick to reserve for byond regardless of its reported internal_tick_usage
 #define TICK_BYOND_RESERVE 2
-#define TICK_LIMIT_RUNNING (max(100 - TICK_BYOND_RESERVE - MAPTICK_LAST_INTERNAL_TICK_USAGE, MAPTICK_MC_MIN_RESERVE))
+/// Tick limit while running normally
+#define TICK_LIMIT_RUNNING (max(TICK_LIMIT_RUNNING_BACKGROUND, MAPTICK_MC_MIN_RESERVE))
+/// Tick limit for background things, strictly obeys the internal_tick_usage metric
+#define TICK_LIMIT_RUNNING_BACKGROUND (100 - TICK_BYOND_RESERVE - MAPTICK_LAST_INTERNAL_TICK_USAGE)
```
Comment (MrStonedOne) [inline code/__DEFINES/_tick.dm:10]: Kill this, it broke things at /tg/ as these things will basically never run above 20 pop.

File: code/controllers/master.dm [hunk 3]
```diff
@@ -424,140 +422,139 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
...
-				tick_precentage = tick_remaining
 
-			tick_precentage = max(tick_precentage*0.5, tick_precentage-queue_node.tick_overrun)
+		//Checks if we're in background calculation mode yet and sets us to it if we are
+		if (!bg_calc && (queue_node_flags & SS_BACKGROUND))
+			current_tick_budget = queue_priority_count_bg
+			bg_calc = TRUE
 
-			current_ticklimit = round(TICK_USAGE + tick_precentage)
+		tick_remaining = ((bg_calc) ? (TICK_LIMIT_RUNNING_BACKGROUND) : (TICK_LIMIT_RUNNING)) - TICK_USAGE
```
Comment (MrStonedOne) [inline code/controllers/master.dm:475]: kill the background stuff here. Im pretty sure this is why things broke at /tg/.

**Review by MrStonedOne — CHANGES_REQUESTED** [review]
we duplicated a little bit of work: #5787

**Comment by TiviPlus** [comment]
Closing in favor of 5787 
because I dont have time to finish this

**Pull request closed without merging**